# Laboratorio 5: Clasificación 🤗

<center><strong>MDS7202: Laboratorio de Programación Científica para Ciencia de Datos - Primavera 2025</strong></center>


### Cuerpo Docente:

- Profesores: Diego Cortez, Gabriel Iturra
- Auxiliares: Melanie Peña, Valentina Rojas
- Ayudantes: Nicolás Cabello, Cristopher Urbina

### Equipo: SUPER IMPORTANTE - notebooks sin nombre no serán revisados

- Nombre de alumno 1: Sebastián Rubina
- Nombre de alumno 2: Maximiliano Farren

### **Link de repositorio de GitHub:** [Repositorio](https://github.com/MacFarren/Lab-Farren-Rubina)

### Temas a tratar
- Clasificación en problemas desbalanceados
- Lightgbm y xgboost
- Pipelines

## Reglas:

- **Grupos de 2 personas**
- Fecha de entrega: Entregas Martes a las 23:59.
- Instrucciones del lab el viernes a las 16:15 en formato online. Asistencia no es obligatoria, pero se recomienda **fuertemente** asistir.
- <u>Prohibidas las copias</u>. Cualquier intento de copia será debidamente penalizado con el reglamento de la escuela.
- Tienen que subir el laboratorio a u-cursos y a su repositorio de github. Labs que no estén en u-cursos no serán revisados. Recuerden que el repositorio también tiene nota.
- Cualquier duda fuera del horario de clases al foro. Mensajes al equipo docente serán respondidos por este medio.
- Pueden usar cualquier material del curso que estimen conveniente.





### Objetivos principales del laboratorio
- Comprender cómo trabajar con problemas de clasificación con clases desbalanceadas.
- Aplicar los modelos lightgbm y xgboost.
- Practicar Pipelines

# Parte Teórica [12 puntos]




1. Explique cuál es la diferencia entre los datos de entrenamiento y validación. [1 punto]

2. Explique cuál es el principal desafío al trabajar problemas de clasificación con data no supervisada. [1 punto]

3. Explique en **sus palabras** qué es la matriz de confusión y para qué se utiliza. [1 puntos]

4. Escriba la fórmula de las siguientes métricas y explique con **sus palabras** cómo se interpretan. [1 punto cada uno]

  * Accuracy
  * Precision
  * Recall
  * F1 score

5. Explique qué métrica recomendaría para los siguientes contextos de clasificación. [1 punto cada uno]

  * Mantenimiento predictivo de fallas de maquinaria pesada en la industria minera.  
  * Detección de enfermedades altamente contagiosas.
  * Aprobación de créditos de alto riesgo.
  * Detección de crímenes.

6. Explique qué es la calibración de modelos y para qué se usa. [1 punto]


**Respuesta:**

*Escriba su respuesta aquí*

P1: Los datos de entrenamiento son la información que se utiliza para ajustar los parámetros del modelo y se utilizan para aprender patrones a partir de ejemplos etiquetados. Los datos de validación son información que se guarda para permanecer al margen del entrenamiento y que son los que se utilizan para las decisiones de modelado que se toman sin sesgo, como pueden ser: seleccionar hiperparámetros, escoger el umbral de la decisión, comparar algoritmos, detectar sobreajuste.

P2: El principal desafío de trabajar problemas de clasificación con data no supervisada es que no existe una "verdad" de referencia, debido a que es necesario elegir grupos o forma de las particiones sin conocer la forma y sin saber si realemente se corresponde a las clases "reales", de esta forma los clusters pueden no alinearse a las categorías relevantes. Además su evaluación es imprecisa, pues se utilizan medidas internas como la silueta.

P3: La matriz de confusión es una representación tabular que sintetiza, por cada clase, las predicciones que el modelo realiza correctamente y las predicciones que efectúa erróneamente: Verdaderos Positivos (TP), Falsos Positivos (FP), Verdaderos Negativos (TN), y Falsos Negativos (FN). La matriz de confusión permite determinar cuál es el tipo de error más cometido por el modelo, permite calcular métricas derivadas de la matriz de confusión, como la precision, el recall, el F1 score, especificidad, y permite ajustar umbrales y costes en función de la criticidad de FP y FN.


P4:
- Accuracy: (TP + TN) / (TP + TN + FP + FN), es la proporción de aciertos globales, sin embargo, con clases desbalanceadas, debido a que un modelo que predice siempre la clase mayoritaria obtiene un accuracy alta. Mientras más alta sea la accuracy es mejor, en caso de clases balanceadas, si no lo está, no sirve tanto ya que puede ser engañosa. Sensible al umbral.

- Precision: (TP / TP + FP), es la correcta predicción de predicciones positivas, cuantas son correctas, alta precisión implica pocos falsos positivos. Se utiliza cuando los falsos positivos son costosos, al aumentar el recall puede bajar y es sensible al umbral. Reportar macro recall si hay desbalance

- Recall: (TP / (TP + FN), se preocupa de los positivos reales, revisa cuantos detecta el modelo. Un alto recall indica que se cuentan con pocos falsos negativos. Util en medicina, donde es muy costoso tener un falso negativo, por ejemplo en cancer. Es sensible al umbral, donde mover el umbral hacia abajo puede subir el recal. Reportar macro recall si hay desbalance

- F1 Score: 2 * (Precision * recall / (precision + recall), es la media armónica entre precisión y recall, es útil cuando se necesita equilibrar ambas clases y hay desbalance. Evita falsos positivos y falsos negativos, sin embargo, al no incorporar los verdaderos negativos, puede ignorar la clase mayoritaria.

P5:

- Mantenimiento predictivo de fallas de maquinaria pesada en la industria minera: Recall, ya que perder una falla implica grandes costos, como puede ser un accidente o maquinaria sin poder utilizarce y perder dinero o tiempo.

- Detección de enfermedades altamente contagiosas: Recall, ya que omitir un falso negaetivo es muy costoso, ya que decirle a una persona que no tiene cancer cuando en verdad si lo tiene. Es mejor decir que si tenia cancer y hacerle más estudios y descartarlo después a decirle que no tenía

- Aprobación de créditos de alto riesgo: Si lo que se busca es aprobar créditos, se recomienda utilizar precision de “aprobar”, ya que se minimizan los falsos positivos, es decir, que se minimiza aprobar a quien no pagará.
Si lo que es positivo es si será moroso, se prioriza Recall de “moroso”, ya que se minimizan los falsos negativos, ya que es importante conceder a quien no pagará.

- Detección de crímenes: Recall, con esto se reducen los falso negativos o no detectar un crimen, nuevamente debido a lo complejo de no detectarlo. Es mejor equivocarse diciendo que si era y que realmente no lo era.

P6: La calibración de modelos ajusta las probabilidades predichas para que reflejen las frecuencias reales de los datos. Se utiliza para tomar desiciones basadas en las probabilidades, donde por ejemplo, es posible fijar umbrales coherentes con la data y priorizar, tener una mayor confianza en lo que se está haciendo, además para combinar modelos y mejorar interpretabilidad de los usuarios.

# Parte práctica [48 puntos]


<p align="center">
  <img src="https://drive.google.com/uc?export=view&id=1BnO4tyh3vM2P199Ec9s3JjngQ4qQ9seP"
" width="300">
</p>


Tras el trágico despido de la mítica mascota de Maipú, Renacín decide adentrarse como consultor en el mercado futbolero, el cuál (para variar...) está cargado en especulaciones.

Como su principal tarea será asesorar a los directivos de los clubes sobre cuál jugador comprar y cuál no, Renacín desea generar modelos predictivos que evaluén distintas características de los jugadores; todo con el fin de tomar decisiones concretas basadas en los datos.

Sin embargo, su condición de corporeo le impidió tomar la versión anterior de MDS7202, por lo que este motivo Renacín contrata a su equipo para lograr su objetivo final. Dado que aún tiene fuertes vínculos con la dirección de deportes de la municipalidad, el corporeo le entrega base de datos con las estadísticas de cada jugador para que su equipo empieze a trabajar ya con un dataset listo para ser usado.


**Los Datos**

Para este laboratorio deberán trabajar con el csv `statsplayers.csv`, donde deberán aplicar algoritmos de aprendizaje supervisado de clasificación en base a características que describen de jugadores de fútbol.

Para comenzar cargue el dataset señalado y a continuación vea el reporte **`Player_Stats_Report.html`** (adjunto en la carpeta del enunciado) que describe las características principales del `DataFrame`.

In [1]:
## Si usted está utilizando Colabolatory le puede ser útil este código para cargar los archivos.
#try:
#    from google.colab import drive
#    drive.mount("/content/drive")
#    path = 'Dirección donde tiene los archivos en el Drive'
#except:
#    print('Ignorando conexión drive-colab')

In [ ]:
#from google.colab import files
#mport pandas as pd


KeyboardInterrupt



In [ ]:
import numpy as np
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import StandardScaler, MinMaxScaler, OneHotEncoder
from sklearn.model_selection import train_test_split
from xgboost import XGBClassifier
from lightgbm import LGBMClassifier
from sklearn.pipeline import Pipeline
from sklearn.metrics import (accuracy_score, precision_score, recall_score, classification_report, confusion_matrix, f1_score)
from sklearn.linear_model import LogisticRegression


In [ ]:
# Carga de datos
# Subir archivo para colab
#uploaded = files.upload()

# Cargar los archivos
df = pd.read_csv("stats_players.csv")

Saving stats_players.csv to stats_players (1).csv


In [ ]:
df

Name          Nationality National_Position Club_Position  \
0      Cristiano Ronaldo             Portugal                LS            LW   
1           Lionel Messi            Argentina                RW            RW   
2                 Neymar               Brazil                LW            LW   
3            Luis Suárez              Uruguay                LS            ST   
4           Manuel Neuer              Germany                GK            GK   
...                  ...                  ...               ...           ...   
17583        Adam Dunbar  Republic of Ireland               NaN           Sub   
17584       Dylan McGoey  Republic of Ireland               NaN           Sub   
17585    Tommy Ouldridge              England               NaN           Res   
17586         Mark Foden             Scotland               NaN           Sub   
17587   Barry Richardson              England               NaN           Sub   

       Height  Weight Preffered_Foot  Age        Work_Rate  Weak_foot  ...  \
0         185      80          Right   32       High / Low          4  ...   
1         170      72           Left   29  Medium / Medium          4  ...   
2         174      68          Right   25    High / Medium          5  ...   
3         182      85          Right   30    High / Medium          4  ...   
4         193      92          Right   31  Medium / Medium          4  ...   
...       ...     ...            ...  ...              ...        ...  ...   
17583     183      82          Right   19  Medium / Medium          1  ...   
17584     185      80          Right   19  Medium / Medium          2  ...   
17585     173      61          Right   18    High / Medium          2  ...   
17586     180      80          Right   21  Medium / Medium          3  ...   
17587     185      77          Right   47  Medium / Medium          2  ...   

       Agility  Jumping  Heading  Shot_Power  Finishing  Long_Shots  Curve  \
0           90       95       85          92         93          90     81   
1           90       68       71          85         95          88     89   
2           96       61       62          78         89          77     79   
3           86       69       77          87         94          86     86   
4           52       78       25          25         13          16     14   
...        ...      ...      ...         ...        ...         ...    ...   
17583       27       56       14          16         13          13     11   
17584       28       53       12          17         12          11     12   
17585       54       61       41          44         28          42     35   
17586       34       48       15          23         14          12     13   
17587       38       51       12          13         11          16     12   

       Freekick_Accuracy  Penalties  Volleys  
0                     76         85       88  
1                     90         74       85  
2                     84         81       83  
3                     84         85       88  
4                     11         47       11  
...                  ...        ...      ...  
17583                 13         15       12  
17584                 13         16       12  
17585                 36         42       37  
17586                 12         24       12  
17587                 11         22       12  

[17588 rows x 39 columns]

## 1. Predicción de Seleccionados Nacionales [14 puntos]

<p align="center">
  <img src="https://www.futuro.cl/wp-content/uploads/2016/06/chile-argentina-meme-12.jpg" width="300">
</p>



### 1.1 Preprocesamiento [5 puntos]

Tareas:

1. Genere los labels para la clasificación binaria en una variable llamada `label`. Para esto, trabaje sobre el atributo `National_Position` suponiendo que los valores nulos son jugadores no seleccionados para representar a su país. [Sin puntaje]

2. Hecho esto, ¿cuántos se tienen ejemplos por cada clase? Comente lo que observa. [1 punto]

3. Genere un `ColumnTransformer` en donde especifique las transformaciones que hay que realizar para cada columna (por ejemplo StandarScaler, MinMaxScaler, OneHotEncoder, etc...) para que puedan ser utilizadas correctamente por el modelo predictivo y guárdelo una variable llamada `col_transformer`. [2 puntos]

4. Comente y justifique las transformaciones elegidas sobre cada una de las variables (para esto utilice el material `Player_Stats_Report.html` que viene en el zip del lab), al igual que las transformaciones aplicadas. [2 puntos]

**Respuesta:**

In [ ]:
#Se genera label sobre el atributo National_Position, suponiendo que los valores nulos son jugadores no seleccionados para representar a su país
df['label'] = df['National_Position'].apply(lambda x: 0 if pd.isna(x) else 1)

In [ ]:
display(df['label'].value_counts())

label
0    16513
1     1075
Name: count, dtype: int64

Label tiene una gran cantidad de datos, es decir, asumiendo que son jugadores que no se han llamado a la selección, solo un 6.5% de los jugadores son llamados a jugar por su seleccion  (solo 1075 jugadores, de un total de 16513), lo que es un número pequeño, pero realista, asumiendo el nivel competitivo del deporte.

In [ ]:
# Definir las transformaciones para cada tipo de columna
# Características numéricas: StandardScaler
# Características categóricas: OneHotEncoder

# Identificar columnas numéricas y categóricas (excluyendo 'label' y 'National_Position')
columnas_numericas = df.select_dtypes(include=np.number).columns.tolist()
columnas_numericas.remove('label')

columnas_categoricas = df.select_dtypes(include='object').columns.tolist()
columnas_categoricas.remove('Name')
columnas_categoricas.remove('Nationality')
columnas_categoricas.remove('Club_Position') # Club_Position se usará en la parte 2
columnas_categoricas.remove('National_Position')

# Crear el ColumnTransformer
col_transformer = ColumnTransformer(
    transformers=[
        ('num', StandardScaler(), columnas_numericas), # Usando StandardScaler para características numéricas
        ('cat', OneHotEncoder(handle_unknown='ignore'), columnas_categoricas) # Usando OneHotEncoder para características categóricas
    ],
    remainder='passthrough' # Mantener otras columnas para uso posterior
)

En un principio se identifican los atributos numericos, por otro lado las categóricas. Para las variables numéricas se decidió aplicar StandardScaler, ya que centra cada característica en media cero y la escala a una desviación estándar unitaria. Ya que con esto se puede trabajar con los datos a una misma escala. Para las categóricas, se eligió el OneHotEncoder con el parámetro handle_unknown='ignore'. Esto convierte cada categoría en una representación binaria sin imponer un orden artificial y, al mismo tiempo, evita errores si durante la fase de prueba aparecen categorías nuevas no vistas en el entrenamiento.

La variable label se mantuvo fuera porque corresponde al objetivo del modelo y procesarla junto con las características produciría data leakage.

La columna Name se descartó debido a su alta cardinalidad y su carácter de identificador único, que no aporta información generalizable y solo incrementaría la dimensionalidad al utilizar OneHotEncoder. Lo mismo sucede con Nationality, que posee más de 160 nacionalidades distintas.

La variable Club_Position no se modificó debido a que después se realiza un tratamiento distinto.

National_Position se elimina debido a que contiene una gran cantidad de datos faltantes

### 1.2 Entrenamiento [3 puntos]

Ahora, vamos a entrenar los pipelines generados en los pasos anteriores. Para esto, debe realizar las siguientes tareas:

1. Separe los datos de entrenamiento en un conjunto de entrenamiento y de prueba  (la proporción queda a su juicio). En este paso, seleccione los ejemplos de forma aleatoria e intente mantener la distribución original de labels de cada clase en los conjuntos de prueba/entrenamiento. (vea la documentación de `train_test_split`). [1 puntos]


2. Defina un pipeline llamado `pipeline_xgboost` y otro llamado `pipeline_lightgbm`. Estos pipelines deben tener el mismo ColumnTransformer definido en la sección de preprocesamiento, pero deben variar los clasificadores de acuerdo al nombre de cada pipeline. [1 puntos]

3. Entrene los pipelines. [1 punto]



**Respuesta:**

In [ ]:
#Armar el set de numéricas + categóricas, además de agregar 'Club_Position'
feature_cols = columnas_numericas + columnas_categoricas
X = df[feature_cols].copy()
y = df['label'].copy()

# 2) Split estratificado para mantener proporción de clases
X_train, X_test, y_train, y_test = train_test_split(
    X, y,
    test_size=0.2,
    random_state=42,
    stratify=y
)

print("Tamaño train:", X_train.shape, "| Tamaño test:", X_test.shape)
print("Distribución train:\n", y_train.value_counts(normalize=True))
print("Distribución test:\n", y_test.value_counts(normalize=True))
print("Ojo, el 35 son las columnas de cada subdf")


Tamaño train: (14070, 35) | Tamaño test: (3518, 35)
Distribución train:
 label
0    0.938877
1    0.061123
Name: proportion, dtype: float64
Distribución test:
 label
0    0.938886
1    0.061114
Name: proportion, dtype: float64
Ojo, el 35 son las columnas de cada subdf


In [ ]:
# Calcular ratio en el set de entrenamiento
n_pos = y_train.sum()
n_neg = len(y_train) - n_pos
ratio = n_neg / n_pos
ratio

15.36046511627907

Se da cuenta que se cuenta con data desbalanceada, por lo que se le asignarán los pesos calculados anteriormente


In [ ]:
# Pipeline con XGBoost
pipeline_xgboost = Pipeline(steps=[
    ('prep', col_transformer),
    ('model', XGBClassifier(
        n_estimators=300,
        learning_rate=0.1,
        max_depth=6,
        subsample=0.8,
        colsample_bytree=0.8,
        random_state=42,
        n_jobs=-1,
        eval_metric='logloss',
        scale_pos_weight=ratio #ratio calculado anteriormente
    ))
])

# Pipeline con LightGBM
pipeline_lightgbm = Pipeline(steps=[
    ('prep', col_transformer),
    ('model', LGBMClassifier(
        n_estimators=300,
        learning_rate=0.1,
        subsample=0.8,
        colsample_bytree=0.8,
        random_state=42,
        n_jobs=-1,
        is_unbalance=True,
        verbose=-1
    ))
])

In [ ]:
# Entrenamiento
pipeline_xgboost.fit(X_train, y_train)
pipeline_lightgbm.fit(X_train, y_train)

Pipeline(steps=[('prep',
                 ColumnTransformer(remainder='passthrough',
                                   transformers=[('num', StandardScaler(),
                                                  ['Height', 'Weight', 'Age',
                                                   'Weak_foot', 'Skill_Moves',
                                                   'Ball_Control', 'Dribbling',
                                                   'Marking', 'Sliding_Tackle',
                                                   'Standing_Tackle',
                                                   'Aggression', 'Reactions',
                                                   'Interceptions', 'Vision',
                                                   'Composure', 'Crossing',
                                                   'Short_Pass', 'Long_Pass',
                                                   'Acceleration', 'Speed',
                                                   'Stamina', 'Strength',
                                                   'Balance', 'Agility',
                                                   'Jumping', 'Heading',
                                                   'Shot_Power', 'Finishing',
                                                   'Long_Shots', 'Curve', ...]),
                                                 ('cat',
                                                  OneHotEncoder(handle_unknown='ignore'),
                                                  ['Preffered_Foot',
                                                   'Work_Rate'])])),
                ('model',
                 LGBMClassifier(colsample_bytree=0.8, is_unbalance=True,
                                n_estimators=300, n_jobs=-1, random_state=42,
                                subsample=0.8, verbose=-1))])

### 1.3 Resultados [6 puntos]

1. Calcule las métricas accuracy, precisión y recall de la clase positiva (la que indica que un jugador es seleccionado nacional) para evaluar el rendimiento de los distintos modelos. Verifique sus resultados usando `classification_report`. [2 puntos]

2. Explique qué implican los valores de accuracy, precisión y recall de la clase positiva (la que indica que un jugador es seleccionado nacional) y cómo influye la cantidad de ejemplos por clase en los resultados obtenidos. [2 puntos]

3. Explique qué métrica le parece más adecuada y concluya qué modelo tiene un mejor desempeño. [2 puntos]

**Respuesta:**

In [ ]:
def evaluar_modelo(nombre, pipeline, X_test, y_test):
    # Predicción con el pipeline
    y_pred = pipeline.predict(X_test)

    # Métricas solicitadas (clase positiva = 1)
    acc  = accuracy_score(y_test, y_pred)
    prec = precision_score(y_test, y_pred, pos_label=1, zero_division=0)
    rec  = recall_score(y_test, y_pred, pos_label=1, zero_division=0)

    print(f"{nombre}")
    print(f"Accuracy:  {acc:.4f}")
    print(f"Precisión (clase 1): {prec:.4f}")
    print(f"Recall    (clase 1): {rec:.4f}\n")

    # Verificación con classification_report
    print("Reporte de clasificacion:")
    print(classification_report(y_test, y_pred, digits=3))
    print("Matriz de confusión:")
    print(confusion_matrix(y_test, y_pred))
    print("-"*50, "\n")

# Evaluar ambos modelos
evaluar_modelo("XGBoost",  pipeline_xgboost,  X_test, y_test)
evaluar_modelo("LightGBM", pipeline_lightgbm, X_test, y_test)


XGBoost
Accuracy:  0.9250
Precisión (clase 1): 0.3497
Recall    (clase 1): 0.2651

Reporte de clasificacion:
              precision    recall  f1-score   support

           0      0.953     0.968     0.960      3303
           1      0.350     0.265     0.302       215

    accuracy                          0.925      3518
   macro avg      0.651     0.617     0.631      3518
weighted avg      0.916     0.925     0.920      3518

Matriz de confusión:
[[3197  106]
 [ 158   57]]
-------------------------------------------------- 

LightGBM
Accuracy:  0.9284
Precisión (clase 1): 0.3851
Recall    (clase 1): 0.2884

Reporte de clasificacion:
              precision    recall  f1-score   support

           0      0.954     0.970     0.962      3303
           1      0.385     0.288     0.330       215

    accuracy                          0.928      3518
   macro avg      0.670     0.629     0.646      3518
weighted avg      0.920     0.928     0.924      3518

Matriz de confusión:
[[320

Para poder realizar de manera correcta el entrenamiento, es que se le ha cambiado la proporción a XGBoost, para intentar manejar el desbalance de clases positivas versus negativas, además de agregarle         is_unbalance=True al segundo modelo, ajustando el umbral de decisión de validación, sin embargo, los resultados muestran el efecto de trabajar con la variable objetivo desbalanceada.

Accuracy, o exactitud, es la proporción de aciertos (TP+TN)/Total, se mantiene alta porque está dominada por los verdaderos negativos de la clase mayoritaria, aun corrigiendo el desbalance en el entrenamiento, sigue reflejando principalmente el buen desempeño sobre los “no seleccionados” y por sí sola no captura la capacidad del modelo para encontrar la clase positiva.

La exactitud indica que tan correctas son las predicciones (TP/(TP+FP), donde al cambiar el umbral para aumentar cobertura, la precisión tiende a descender debido a que aceptamos más falsos positivos.

El Recall mide la cobertura sobre los verdaderos seleccionados, en donde esta es la medida más sensible al desbalance, debido a que con pocos ejemplos, se dejan escapar los falsos negativos.



Debido al claro desbalance, es que las métricas no son suficientes para poder predecir de manera correcta, sin embargo, el modelo que presenta el mejor equilibrio entre desempeño global y capacidad de detección de la clase positiva es LightGBM. En particular, mantiene una accuracy comparable a XGBoost (≈0,928 frente a ≈0,925), pero lo supera en los otros indicadores para el objetivo del estudio, posee una precisión de la clase positiva superior (≈0,385 vs. ≈0,350) y un recall también mayor (≈0,288 vs. ≈0,265), lo que implica identificar una mayor proporción de jugadores efectivamente seleccionados con menor proporción de falsos positivos que su contraparte. Dado el desbalance entre clases, el accuracy puede no ser la mejor opción, la combinación de mayor recall con una precisión no disminuida confirma que LightGBM ofrece la mejor relación señal-ruido sobre la clase de interés.

## 2. Predicción de posiciones de jugadores [4 puntos]

<p align="center">
  <img src="https://pbs.twimg.com/media/E1rfA1aWEAYU6Ny.jpg" width="300">
</p>

En una nueva jornada de desmesuradas transacciones deportivas, Renacín escuchó a sus colegas discutir acerca de que el precio de cada jugador depende en gran medida de la posición en la cancha en la que juega. Y además, que hay bastantes jugadores nuevos que no tienen muy claro en que posición verdaderamente brillarían, por lo que actualmente puede que actualmente estén jugando en posiciones sub-optimas.

Viendo que los resultados del primer análisis no son tan esperanzadores, el corporeo los comanda a cambiar su tarea: ahora, les solicita que construyan un clasificador enfocado en predecir la mejor posición de los jugadores en la cancha según sus características.

Para lograr esto, primero, les pide que etiqueten de la siguiente manera los valores que aparecen en el atributo `Club_Position`, pidiendo que agrupen los valores en los siguientes grupos:

**Nota**:  Renacín les recalca que **no deben utilizar los valores ```Sub``` y ```Res``` de esta columna**.

```python
ataque = ['ST', 'CF']
central_ataque = ['RW', 'CAM', 'LW']
central = ['RM', 'CM', 'LM']
central_defensa = ['RWB', 'CDM', 'LWB']
defensa = ['RB', 'CB', 'LB']
arquero = ['GK']
```

La elección del clasificador se justificar en base a la siguiente [guía](https://scikit-learn.org/stable/tutorial/machine_learning_map/index.html) y se deben comentar los resultados obtenidos en la clasificación.

**Tareas:** [1 punto por tarea]

1. En un nuevo dataframe, aplique las etiquetas descritas anteriormente en cada uno de los valores señalados en esta sección y guárdelos en la variable `label`.
2. Cuente cuántos por clase quedan.
3. Entrene el nuevo pipeline y ejecute una evaluación de este.  
4. Comente los resultados obtenidos.

**Respuesta:**

In [ ]:
# Creamos un nuevo data frame
df_pos = df.copy()

# Cramos lo solicitado según atributos
grupo_map = {
    'ataque': {'ST', 'CF'},
    'central_ataque': {'RW', 'CAM', 'LW'},
    'central': {'RM', 'CM', 'LM'},
    'central_defensa': {'RWB', 'CDM', 'LWB'},
    'defensa': {'RB', 'CB', 'LB'},
    'arquero': {'GK'}
}

# se eliminan "sub" y "res"
mask_valid = df_pos['Club_Position'].notna() & (~df_pos['Club_Position'].isin(['Sub', 'Res']))
df_pos = df_pos.loc[mask_valid].copy()

# Función de mapeo
def map_pos(value: str) -> str:
    for g, posiciones in grupo_map.items():
        if value in posiciones:
            return g
    return np.nan

df_pos['label'] = df_pos['Club_Position'].apply(map_pos)

# Contamos por clase para revisar no tener el mismo error de antes
print("Conteo por clase (label):")
print(df_pos['label'].value_counts())
print()



Conteo por clase (label):
label
defensa            1180
central             907
arquero             632
central_ataque      581
ataque              430
central_defensa     209
Name: count, dtype: int64



In [ ]:
# Definir columnas, al igual que lo hicimos anteriormente, excluimos lo que no no queremos que esté en el modelo

columnas_numericas_pos = df_pos.select_dtypes(include=np.number).columns.tolist()
if 'label' in columnas_numericas_pos:
    columnas_numericas_pos.remove('label')

columnas_categoricas_pos = df_pos.select_dtypes(include='object').columns.tolist()
for col_drop in ['label', 'Club_Position', 'National_Position', 'Name', 'Nationality']:
    if col_drop in columnas_categoricas_pos:
        columnas_categoricas_pos.remove(col_drop)

#Creamos el column transformer pero ahora para posicion
col_transformer_pos = ColumnTransformer(
    transformers=[
        ('num', StandardScaler(), columnas_numericas_pos),
        ('cat', OneHotEncoder(handle_unknown='ignore'), columnas_categoricas_pos),
    ],
    remainder='drop'
)

# Preparar X, y y split multiclase
X_pos = df_pos.drop(columns=['label', 'Club_Position', 'National_Position', 'Name', 'Nationality'], errors='ignore')
y_pos = df_pos['label'].astype('category')

# Eliminar filas con NaN en 'label' antes de dividir los datos
df_pos_cleaned = df_pos.dropna(subset=['label']).copy()
X_pos_cleaned = df_pos_cleaned.drop(columns=['label', 'Club_Position', 'National_Position', 'Name', 'Nationality'], errors='ignore')
y_pos_cleaned = df_pos_cleaned['label'].astype('category')


X_train, X_test, y_train, y_test = train_test_split(
    X_pos_cleaned, y_pos_cleaned, test_size=0.2, random_state=42, stratify=y_pos_cleaned
)

# Definir el pipeline para LogisticRegression multinomial, según el cheat sheet
pipeline_pos = Pipeline(steps=[
    ('prep', col_transformer_pos),
    ('model', LogisticRegression(
        solver='saga', # multi_class='multinomial' is default for saga, removing deprecated parameter
        max_iter=2000, n_jobs=-1, random_state=42
    ))
])

# Entrenar modelo
pipeline_pos.fit(X_train, y_train)

# Evaluar modelos
y_pred = pipeline_pos.predict(X_test)

acc  = accuracy_score(y_test, y_pred)
f1m  = f1_score(y_test, y_pred, average='macro')  # macro-F1: justo con clases desbalanceadas
print(f"Accuracy:  {acc:.4f}")
print(f"Macro-F1:  {f1m:.4f}\n")

print("Reporte de clasificación:")
print(classification_report(y_test, y_pred, target_names=sorted(y_pos_cleaned.unique().tolist()), digits=3))

print("Matriz de confusión (orden etiquetas):", sorted(y_pos_cleaned.unique().tolist()))
print(confusion_matrix(y_test, y_pred, labels=sorted(y_pos_cleaned.unique().tolist())))

Accuracy:  0.7449
Macro-F1:  0.6629

Reporte de clasificación:
                 precision    recall  f1-score   support

        arquero      1.000     1.000     1.000       126
         ataque      0.771     0.860     0.813        86
        central      0.584     0.610     0.597       182
 central_ataque      0.494     0.362     0.418       116
central_defensa      0.400     0.190     0.258        42
        defensa      0.834     0.958     0.892       236

       accuracy                          0.745       788
      macro avg      0.681     0.663     0.663       788
   weighted avg      0.723     0.745     0.729       788

Matriz de confusión (orden etiquetas): ['arquero', 'ataque', 'central', 'central_ataque', 'central_defensa', 'defensa']
[[126   0   0   0   0   0]
 [  0  74   8   4   0   0]
 [  0   8 111  36   7  20]
 [  0  13  56  42   2   3]
 [  0   1   9   2   8  22]
 [  0   0   6   1   3 226]]


Notamos el modelo es aceptable (accuracy = 0,745) y razonable en términos de equilibrio entre clases (macro-F1 = 0,663), sin embargo, la matriz de confusión y los indicadores por clase evidencian diferencias marcadas de dificultad según la posición.


Los roles extremos se distinguen de manera correctar (GK, defensa, ataque), sin embargo, le cuesta separar sub-roles del mediocampo, donde el perfil de atributos es más parecido y las etiquetas agregadas introducen zonas limítrofes.
Las clases arquero y defensa son las más separables, arquero alcanza un desempeño perfecto (P=R=F1=1,00), consistente con atributos muy distintivos del puesto, y defensa logra resultados altos (P=0,834; R=0,958; F1=0,892), lo que sugiere que el modelo capta bien los rasgos defensivos. ataque también presenta un rendimiento sólido (F1=0,813) con buen recall (0,860). En contraste, las clases del medio muestran mayor solapamiento, donde central queda en un nivel intermedio (F1=0,597) y central_ataque (F1=0,418; R=0,362) y central_defensa (F1=0,258; R=0,190) exhiben el peor comportamiento. La confusión se concentra, principalmente, en central_ataque que es reconocido como central (56 casos) y central_defensa como defensa (22 casos), además de desvíos de central a central_ataque (36) y central a defensa (20). De esta forma, se nota que los roles contiguos no están suficientemente separadas por las variables actuales o requieren relaciones no lineales para ser captadas.


## 3. Predicciones de Seleccionados Nacionales para el Jere Klein [30 puntos]

<center>
<img src='https://www.radioactiva.cl/wp-content/uploads/2024/04/Jere-Klein-1-768x432.webp' width=500 />

Después de alcanzar la fama como cantante urbano, Jere Klein decide explorar una nueva faceta. Con su amor por el fútbol y convencido de que los artistas urbanos poseen un talento y versatilidad excepcionales, Jere se embarca en un proyecto innovador: desarrollar un sistema de inteligencia artificial capaz de identificar a jugadores que tienen potencial para convertirse en futbolistas profesionales. Su teoría es que muchos artistas del género urbano chileno, con sus habilidades únicas y su disciplina, podrían destacarse también en el deporte. Con este sistema, Jere espera no solo abrir nuevas oportunidades para sus colegas artistas, sino también demostrar la amplia gama de talentos que pueden ofrecer.

### 2.1 ¿Qué modelo de árbol es más de "pana"? [10 puntos]

<center>
<img src='https://64.media.tumblr.com/39189215a7d3d96823cb359f35b44e05/tumblr_psmrhrR3Xw1qf5hjqo4_540.gif' width=300 />


**Tareas**


1. Considerando el la variable llamada `label` creada en la sección 1.1. Para determinar cuál modelo de árbol sería más adecuado para la tarea en cuestión, utilice PyCaret. Este deberá centrarse exclusivamente en modelos de tipo árbol. Jere ha especificado que busca un modelo que tome decisiones rápidamente y que tenga una baja tasa de falsos positivos, ya que planea invertir en estos jugadores. [3 puntos]

Para la comparación, utilice los siguientes modelos:

```python
['et', 'rf', 'dt', 'xgboost', 'lightgbm', 'catboost']
```

2. Explique en brevemente que son los modelos de la siguiente lista `['et', 'rf', 'dt']` y como funcionan. [3 punto]

3. Tras realizar la comparación de modelos, seleccione aquel que muestre el mejor rendimiento en términos de velocidad y precisión, especialmente en la reducción de falsos positivos. Utilice la función `evaluate_model` de PyCaret para revisar y analizar los resultados obtenidos en los siguientes aspectos:

  - **Confusión Matrix**: ¿Cómo se encuentran la tasa de verdaderos positivos y verdaderos negativos?
  - **Threshold**: ¿Es acaso el umbral por defecto del modelo el mejor para las predicciones?
  - **Feature Importance**: ¿Cuáles son las variables con mejor desempeño? ¿A qué podría deberse esto?
  - **Learning Curve**: ¿El modelo presenta algún problema?

  [4 puntos]

**Respuesta**

## Explique en brevemente que son los modelos de la siguiente lista ['et', 'rf', 'dt'] y como funcionan.

Los modelos de la lista son algoritmos basados en árboles de decisión que se utilizan para clasificación y regresión.

Extra Trees es un método de ensamble que crea varios árboles de decisión utilizando todo el conjunto de datos, pero introduce aleatoriedad en la forma en que selecciona los puntos de corte para dividir los nodo. A diferencia de Random Forest, no utiliza bootstrap y elige los umbrales de división de manera completamente aleatoria, lo que lo hace más rápido de entrenar aunque potencialmente menos preciso. La predicción final se obtiene promediando las predicciones de todos los árboles.
Random Forest también es un método ensemble que construye múltiples árboles de decisión, pero lo hace creando muestras bootstrap del conjunto de datos original y seleccionando aleatoriamente un subconjunto de características en cada división. Cada árbol se entrena de forma independiente y la predicción final se determina por votación mayoritaria en clasificación o promedio en regresión. Este método es robusto frente al sobreajuste y generalmente ofrece un buen balance entre precisión y generalización.
Decision Tree es un modelo individual que construye una estructura jerárquica de decisiones dividiendo repetidamente los datos según las características que mejor separan las clases o minimizan el error. Comienza desde la raíz y va creando ramas mediante preguntas binarias sobre las variables hasta llegar a las hojas que contienen las predicciones finales. Es muy interpretable y rápido, pero propenso al sobreajuste si no se controla su complejidad mediante poda o limitación de profundidad.

A la hora de comentar sobre cuando se usa cada uno, et es ideal cuando se necesita entrenar modelos rápidamente y hacer predicciones veloces, especialmente con grandes conjuntos de datos. Es útil para reducir el sobreajuste, manejar muchas características y trabajar con datos ruidosos.

Random Forest es una opción versátil y equilibrada. Funciona bien en la mayoría de los casos sin requerir muchos ajustes, y es útil para medir la importancia de las características. Es la opción más segura al comenzar un proyecto pero un poquito lenta.

Decision Tree es adecuado para modelos simples e interpretables, especialmente cuando la transparencia es crucial. Es ideal para conjuntos de datos pequeños y cuando se requiere velocidad en las predicciones, pero hay que tener cuidado con el sobreajuste.

In [ ]:
#!pip install pycaret
#!pip install --upgrade pycaret

In [ ]:
!pip install pycaret --no-deps
!pip install pandas numpy scikit-learn scipy matplotlib seaborn plotly ipywidgets joblib

In [ ]:
from pycaret.datasets import get_data
from pycaret.classification import *
import os

os.environ["PYCARET_CUSTOM_LOGGING_LEVEL"] = "CRITICAL"

# Continuar código aquí

# Prepare the data for PyCaret
# We use the original df as the label is based on National_Position
data_pycaret = df.copy()

# Setup PyCaret environment
# The target variable is 'label'
# We ignore columns that are not relevant for modeling
# session_id for reproducibility
# fix_imbalance to address the class imbalance
clf1 = setup(data = data_pycaret, target = 'label', session_id = 42,
            ignore_features = ['Name', 'Nationality', 'Club_Position', 'National_Position'],
            fix_imbalance = True,
            log_experiment=True, experiment_name='national_selection_tree_models')

# Compare specified tree-based models
# Use sort='Precision' to prioritize models with lower false positives
best_model = compare_models(
    include = ['et', 'rf', 'dt', 'xgboost', 'lightgbm', 'catboost'],
    sort = 'Precision'
)

print(best_model)

RuntimeError: ('Pycaret only supports python 3.9, 3.10, 3.11. Your actual Python version: ', sys.version_info(major=3, minor=12, micro=11, releaselevel='final', serial=0), 'Please DOWNGRADE your Python version.')

### 2.2 Reducción de dimensionalidad [14 puntos]

<center>
<img src='https://i.kym-cdn.com/photos/images/original/002/258/560/668.gif' width=400 />

A pesar de los resultados obtenidos previamente, el manager de Jere ha solicitado el entrenamiento de un modelo de XGBoost utilizando los datos disponibles. Además, se debe proceder a realizar una reducción de dimensionalidad basada en la importancia de las características.

Para llevar a cabo esta tarea:

1. Inicie entrenando un modelo XGBoost con todas las características disponibles. [2 puntos]

2. Una vez el modelo esté entrenado, evalúe y clasifique las características según su importancia de forma descendente. [2 puntos]

3. Utilice esta clasificación para ejecutar una búsqueda recursiva de eliminación de características, eliminando progresivamente las menos importantes y evaluando el impacto en el desempeño del modelo hasta identificar las N características más críticas. [2 puntos]

4. Con este conjunto reducido de características, entrene un nuevo modelo y evalúe su rendimiento. [2 puntos]

5. Posteriormente, responda a las siguientes preguntas para una comprensión más profunda de los cambios y beneficios:

  - ¿El rendimiento del modelo con las características seleccionadas es similar al del modelo original? ¿Cómo se comparan en términos de precisión y robustez? [2 puntos]
  - ¿Cuáles son los beneficios potenciales de eliminar variables del modelo? Considere factores como la simplificación del modelo, reducción del tiempo de entrenamiento, y mejora en la capacidad de generalización. [2 puntos]
  - Comente si el modelo con menor dimensionalidad es más sencillo de explicar. Explique brevemente por qué la eliminación de ciertas características puede facilitar la comprensión y la explicación del comportamiento del modelo. [2 puntos]

Notar que con esta metodologia buscamos encontrar un punto entermedio entre número de festures y desempeño. por esto, si observa que al aumentar festires el aumento es despreciable, puede no considerar agregar más features a su modelo.

**Respuesta**

### 2.3 Calibración Probabilistica [6 puntos]

<center>
<img src='https://media2.giphy.com/media/l2Je4Ku0Cx292KWv6/200w.gif?cid=6c09b952y0sihtq9tb6sz8j2023x3zxxp3qx1ocgonkpkblj&ep=v1_gifs_search&rid=200w.gif&ct=g' width=400 />

Para lograr modelos más modulares, se recomienda realizar una calibración del modelo entrenado anteriormente, con el objetivo de obtener salidas que reflejen mayor modularidad.

1. Se solicita que utilice un método de calibración que asegure que las probabilidades generadas incrementen de manera monótona. Una métrica ampliamente utilizada para evaluar la precisión de la calibración de un modelo es el Brier Score. Calcule el Brier Score para el modelo tanto antes como después de la calibración. Esto le permitirá realizar una comparación cuantitativa y determinar si la calibración ha mejorado el rendimiento del modelo. Para más información sobre el Brier Score, puede consultar el siguiente enlace: [Scikit-Learn - Brier Score Loss](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.brier_score_loss.html). [3 puntos]

2. Tras la calibración, examine y comente los resultados obtenidos. A su análisis añada una comparación visual de las ideales versus las salidas del modelo original (sin calibrar) y del modelo calibrado. [3 puntos]

**Respuesta**

Mucho éxito!

<center>
<img src="https://i.pinimg.com/originals/55/3d/42/553d42bea9b10e0662a05aa8726fc7f4.gif" width=300>